src/compiler/checker.ts

```diff
@@ -37684,7 +37684,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
     function checkReferenceExpression(expr: Expression, invalidReferenceMessage: DiagnosticMessage, invalidOptionalChainMessage: DiagnosticMessage): boolean {
         // References are combinations of identifiers, parentheses, and property accesses.
-        const node = skipOuterExpressions(expr, OuterExpressionKinds.Assertions | OuterExpressionKinds.Parentheses);
+        const node = skipOuterExpressions(expr, OuterExpressionKinds.Assertions | OuterExpressionKinds.SatisfiesExpressions | OuterExpressionKinds.Parentheses);

```

Comment:
@Andarist 

I added `OuterExpressionKinds.SatisfiesExpressions` to the skipped outer expressions.

I tracked the calls to `checkReferenceExpression` and found that several uses cases had been silently broken.
```ts
let a = 10;

// called in checkPrefixUnaryExpression
--(a satisfies number);
++(a satisfies number);

// called in checkPostfixUnaryExpression
(a satisfies number)++;
(a satisfies number)--;

// called in checkAssignmentOperatorWorker
let b: number;
(b satisfies number) = 10;

// called in checkReferenceAssignment called in checkDestructuringAssignment
let c: number;
[(c satisfies number)] = [10];

let d: number, e = 1;
({ d: (e satisfies number) } = { d: 10 });

// called in checkForInStatement
let x: string = "hello"
for ((x satisfies string) in { a: 10 }) {
  console.log(x)
}

// called in checkForOfStatement
let g = 1
for ((g satisfies number) of [10]) {
  console.log(g)
}
```

Adding `OuterExpressionKinds.SatisfiesExpressions` corrects the error, and I've added the tests to prove it.

Reply:
Great and thorough investigative work 👏